### PR TITLE
add Optimi(more fused_background_optimizer and new function)

### PR DESCRIPTION
**Pull request by sdbds**
https://optimi.benjaminwarner.dev/

New optimizer:
adamw、lion、ranger and stableadamw from Optimi

New function:
Low Precision Training with Kahan Summation(auto use when use above optimizers)
image

Gradient Release(same as fused background pass)
Auto use when use above optimizers.

Fully Decoupled Weight Decay(looks like decoupled lr)

Optimizer Accumulation
[Gradient accumulation](https://pytorch.org/docs/stable/notes/amp_examples.html#gradient-accumulation) reduces training memory by splitting a batch into micro-batches and accumulating micro-batch gradients into the larger batch. [Gradient release](https://optimi.benjaminwarner.dev/gradient_release/) reduces training memory by limiting gradients to one layer at any given time. Optimizer accumulation unifies these two disparate approaches by accumulating gradients directly into optimizer states while performing gradient release.

### CODE REVIEW

**Comment by sdbds** [comment]
The problem now is that the VRAM usage keeps going up after using prepare_for_gradient_release() for hook, not sure about the problem at the moment...

**Comment by feffy380** [comment]
Some of optimi's features do not support fp16. They should not replace the original optimizers.

**Comment by sdbds** [comment]
> Some of optimi's features do not support fp16. They should not replace the original optimizers.

Sure, i will add fp16 checker, thanks

**Comment by feffy380** [comment]
I'm saying it should not be considered a substitute for the original optimizers. Some features do work with fp16 (like low precision training), but others don't.

You don't need an fp16 check, you need to treat the optimi versions as completely separate optimizers, the same way Adamw8bit is considered separate from Adamw
